### PR TITLE
[SPARK-22660][BUILD] Use position() and limit() to fix ambiguity issue in scala-2.12

### PR DESCRIPTION
**Pull request by kellyzly**
…a-2.12 and JDK9

## What changes were proposed in this pull request?
Some compile error after upgrading to scala-2.12
```javascript
spark_source/core/src/main/scala/org/apache/spark/executor/Executor.scala:455: ambiguous reference to overloaded definition, method limit in class ByteBuffer of type (x$1: Int)java.nio.ByteBuffer
method limit in class Buffer of type ()Int
match expected type ?
     val resultSize = serializedDirectResult.limit
error  
```
The limit method was moved from ByteBuffer to the superclass Buffer and it can no longer be called without (). The same reason for position method.

```javascript
/home/zly/prj/oss/jdk9_HOS_SOURCE/spark_source/sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala:427: ambiguous reference to overloaded definition, [error] both method putAll in class Properties of type (x$1: java.util.Map[_, _])Unit [error] and  method putAll in class Hashtable of type (x$1: java.util.Map[_ <: Object, _ <: Object])Unit [error] match argument types (java.util.Map[String,String])
 [error]       props.putAll(outputSerdeProps.toMap.asJava)
 [error]             ^
 ```
This is because the key type is Object instead of String which is unsafe.

## How was this patch tested?

running tests 

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Review by srowen — COMMENTED** [review]
This kind of change is fine as it's backwards-compatible, if it resolves a real compiler issues later. But collect as many of this type of problem in this PR rather than across several.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
Let's fix the title to `[SPARK-22660][BUILD] Use position() and limit() to fix ambiguity issue in scala-2.12 and JDK9` and fill up the PR description BTW.

File: core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala [hunk 1]
```diff
@@ -99,7 +99,7 @@ private[spark] class TorrentBroadcast[T: ClassTag](obj: T, id: Long)
   private def calcChecksum(block: ByteBuffer): Int = {
     val adler = new Adler32()
     if (block.hasArray) {
-      adler.update(block.array, block.arrayOffset + block.position, block.limit - block.position)
+      adler.update(block.array, block.arrayOffset + block.position(), block.limit() - block.position())
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala]: This exceeds 100 character length limit. See the logs in Jenkins tests.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java [hunk 2]
```diff
@@ -169,8 +169,10 @@ public static ByteBuffer allocateDirectBuffer(int size) {
       cleanerField.setAccessible(true);
       long memory = allocateMemory(size);
       ByteBuffer buffer = (ByteBuffer) constructor.newInstance(memory, size);
-      Cleaner cleaner = Cleaner.create(buffer, () -> freeMemory(memory));
-      cleanerField.set(buffer, cleaner);
+//    sun.misc.Cleaner is not used in JDK9(see SPARK-22659), we can use CleanUtil in HADOOP-12760 to solve it
+//    Cleaner cleaner = Cleaner.create(buffer, () -> freeMemory(memory));
+//    cleanerField.set(buffer, cleaner);
```
Comment (srowen) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java]: This change isn't valid and isn't related

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala [hunk 3]
```diff
@@ -412,7 +412,9 @@ case class HiveScriptIOSchema (
     propsMap = propsMap + (serdeConstants.LIST_COLUMN_TYPES -> columnTypesNames)
 
     val properties = new Properties()
-    properties.putAll(propsMap.asJava)
+    // properties.putAll(propsMap.asJava)
+    // see https://github.com/apache/kafka/pull/3647
+    propsMap.foreach{ case (k, v) => properties.put(k, v) }
```
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala]: Nit: space before brace

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala [hunk 4]
```diff
@@ -296,7 +296,8 @@ class KafkaTestUtils(withBrokerProps: Map[String, Object] = Map.empty) extends L
     props.put("replica.socket.timeout.ms", "1500")
     props.put("delete.topic.enable", "true")
     props.put("offsets.topic.num.partitions", "1")
-    props.putAll(withBrokerProps.asJava)
+    // props.putAll(withBrokerProps.asJava)
+    withBrokerProps.foreach{ case (k, v) => props.put(k, v) }
```
Comment (viirya) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala]: Is this related?

Comment (viirya) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala]: https://github.com/scala/bug/issues/10418. I see.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala (hunk 3 above)
Comment (viirya) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala]: I think the direct reference should be https://github.com/scala/bug/issues/10418.

Comment (viirya) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala]: And we should not leave commented code.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala [hunk 5]
```diff
@@ -424,7 +426,9 @@ case class HiveScriptIOSchema (
     recordReaderClass.map { klass =>
       val instance = Utils.classForName(klass).newInstance().asInstanceOf[RecordReader]
       val props = new Properties()
-      props.putAll(outputSerdeProps.toMap.asJava)
+      // props.putAll(outputSerdeProps.toMap.asJava)
+      // see https://github.com/apache/kafka/pull/3647
```
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala]: ditto. I think it is better to remove commented code and refer to https://github.com/scala/bug/issues/10418.

**Comment by srowen** [comment]
Ping @kellyzly

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala (hunk 4 above)
Comment (kellyzly) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala]: yes, the change to the putAll is similar as the solution provided in scala/bug#10418

File: common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java (hunk 2 above)
Comment (kellyzly) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java]: Sure I can remove the comment like " sun.misc.Cleaner is not used in JDK9(see SPARK-22659), we can use CleanUtil in HADOOP-12760 to solve it"
sun.misc.Cleaner has moved to a new location in OpenJDK 9.  If uncommenting code about Cleaner, compile will fail. Cleaner is used to [release memory](http://www.docjar.com/html/api/sun/misc/Cleaner.java.html).  If commenting code about cleaner, there maybe some function lost. Can you give me some suggestions?

Comment (kellyzly) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java]: @srowen ： have deleted changes about Platform.java. So current pr is only for solving problems met in scala-2.12 and can not fix the problem about JDK9.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala (hunk 4 above)
Comment (HyukjinKwon) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala]: Nit: `h{ ` -> `h { `

**Comment by jiangxb1987** [comment]
Please fix the title and also fulfill the PR description.

**Review by srowen — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by kellyzly** [comment]
thanks @HyukjinKwon ,@srowen, @viirya 's review

**Comment by srowen** [comment]
Merged to master
